lib/services/itmstransporter-service.ts

```diff
@@ -0,0 +1,88 @@
+///<reference path="../.d.ts"/>
+"use strict";
+import * as path from "path";
+import * as temp from "temp";
+let md5 = require("cryptojs").Crypto.MD5;
+
+export class ITMSTransporterService implements IITMSTransporterService {
+	private _itmsTransporterPath: string = null;
+
+	constructor(private $childProcess: IChildProcess,
+		private $devicePlatformsConstants: Mobile.IDevicePlatformsConstants,
+		private $errors: IErrors,
+		private $fs: IFileSystem,
+		private $platformService: IPlatformService,
+		private $staticConfig: IStaticConfig,
+		private $sysInfo: ISysInfo) { }
+
+	public upload(appId: string, username: string, password: string, mobileProvisionIdentifier?: string, codeSignIdentity?: string): IFuture<void> {
+		return (() => {
+			temp.track();
+			let itmsTransporterPath = this.getITMSTransporterPath().wait(),
+				ipaFileName = "app.ipa",
+				itmsDirectory = temp.mkdirSync("itms-"),
+				innerDirectory = path.join(itmsDirectory, "mybundle.itmsp"),
+				ipaFileLocation = path.join(innerDirectory, ipaFileName),
+				platform = this.$devicePlatformsConstants.iOS,
+				forDevice = true,
+				iosBuildConfig: IiOSBuildConfig = {
+					buildForDevice: forDevice,
+					mobileProvisionIdentifier: mobileProvisionIdentifier,
+					codeSignIdentity: codeSignIdentity
+				};
+
+			this.$fs.createDirectory(innerDirectory).wait();
+
+			this.$platformService.preparePlatform(platform).wait();
+			this.$platformService.buildPlatform(platform, iosBuildConfig).wait();
+			this.$platformService.copyLastOutput(platform, ipaFileLocation, { isForDevice: forDevice }).wait();
+
+			let ipaFileHash = md5(this.$fs.readFile(ipaFileLocation).wait()),
+				ipaFileSize = this.$fs.getFileSize(ipaFileLocation).wait(),
+				metadata = `<?xml version="1.0" encoding="UTF-8"?>
+<package version="software4.7" xmlns="http://apple.com/itunes/importer">
+    <software_assets apple_id="${appId}">
+        <asset type="bundle">
+            <data_file>
+                <file_name>${ipaFileName}</file_name>
+                <checksum type="md5">${ipaFileHash}</checksum>
+                <size>${ipaFileSize}</size>
+            </data_file>
+        </asset>
+    </software_assets>
+</package>`;
+
+			this.$fs.writeFile(path.join(innerDirectory, "metadata.xml"), metadata).wait();
+
+			this.$childProcess.spawnFromEvent(itmsTransporterPath, ["-m", "upload", "-f", itmsDirectory, "-u", username, "-p", password, "-v", "informational"], "close", { stdio: "inherit" }).wait();
+		}).future<void>()();
+	}
+
+	private getITMSTransporterPath(): IFuture<string> {
+		return (() => {
+			if (!this._itmsTransporterPath) {
+				let sysInfo = this.$sysInfo.getSysInfo(path.join(__dirname, "..", "..", this.$staticConfig.PROJECT_FILE_NAME)).wait(),
```

Comment:
you can pass `this.$staticConfig.pathToPackageJson` instead of constructing the path